lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,354 @@
+static P visitHelper(R (C::*)(P) const);
+
+// Helper for visiting module operations non-recursively.
+// Allows erasing the visited operation during visitation.
+template<typename Visitor>
+static void visit(ModuleOp Module, Visitor visitor) {
+  auto &OpList = getModuleOperations(Module);
+
+  auto Begin = OpList.begin();
+  const auto End = OpList.end();
+
+  while (Begin != End) {
+    using RequestedOpT = decltype(visitHelper(&Visitor::operator()));
+    Operation *const Op = &*Begin++;
+    if constexpr (std::is_same_v<RequestedOpT, Operation *>) {
+      visitor(Op);
+    } else if (auto RequestedOp = mlir::dyn_cast<RequestedOpT>(Op)) {
+      visitor(RequestedOp);
+    }
+  }
+}
+
+static const mlir::DialectRegistry &getDialectRegistry() {
+  static const mlir::DialectRegistry Registry = []() -> mlir::DialectRegistry {
+    mlir::DialectRegistry Registry;
+    // The DLTI dialect is used to express the data layout.
+    Registry.insert<mlir::DLTIDialect>();
+    // All dialects that implement the LLVMImportDialectInterface.
+    mlir::registerAllFromLLVMIRTranslations(Registry);
+    Registry.insert<mlir::clift::CliftDialect>();
+    return Registry;
+  }();
+  return Registry;
+}
+
+static ContextPtr makeContext() {
+  const auto Threading = MLIRContext::Threading::DISABLED;
+  return std::make_unique<MLIRContext>(getDialectRegistry(), Threading);
+}
+
+// Cloning MLIR from one module into another requires first serialising the
+// source and then deserialising it again into the target module. We do this
+// a) when cloning operations from one container to another, and
+// b) when "garbage collecting" types and attributes during remove.
+//
+// Ideally at least for use case a) we could clone from one context into another
+// directly, but that is not supported at this time.
+static OwningModuleRef cloneModuleInto(ModuleOp SourceModule,
+                                       MLIRContext &DestinationContext) {
+  llvm::SmallString<1024> Buffer;
+  llvm::raw_svector_ostream Out(Buffer);
+  mlir::writeBytecodeToFile(SourceModule, Out);
+
+  // Serialised MLIR contains multiple operations and as such must be
+  // deserialised into a block. We use a temporary block stored on stack.
+  mlir::Block OuterBlock;
+
+  const mlir::LogicalResult
+    R = mlir::readBytecodeFile(llvm::MemoryBufferRef(Buffer, "cloneModuleInto"),
+                               &OuterBlock,
+                               mlir::ParserConfig(&DestinationContext));
+  revng_assert(mlir::succeeded(R));
+
+  // The serialised code is known to contain exactly one ModuleOp operation.
+  auto &Operations = OuterBlock.getOperations();
+  revng_assert(Operations.size() == 1);
+  auto DestinationModule = mlir::cast<ModuleOp>(Operations.front());
+
+#ifndef NDEBUG
+  const size_t SrcOpsCount = getModuleOperations(SourceModule).size();
+  const size_t NewOpsCount = getModuleOperations(DestinationModule).size();
+  revng_assert(NewOpsCount == SrcOpsCount);
+#endif
```

Comment:
Nice. Sometimes, for expensive checks, we wrap the checks in

```c++
if (VerifyLog.isEnabled()) {
  // ...
}
```
Then you can enable them passing `--debug-log=verify`, which is used in testing and is useful if you get weird crashes. In future the flag will become something like `--extra-checks`, for now we abuse loggers.